### PR TITLE
Regex: add some simple regex method

### PR DESCRIPTION
**Pull request by Liaojinghui**
https://github.com/neo-project/neo/issues/2922

Make it easier to parse string maybe.

adding methods:

startwith, endwith, contains, substring, indexof

### CODE REVIEW

**Review by cschuchardt88 — APPROVED** [review]
thx for getting it going 😃

**Review by roman-khimov — COMMENTED** [review]
Seems to me all of these can be done with `StdLib` already, there is `MemorySearch` there. `SUBSTR` opcode is available as well.

**Comment by Liaojinghui** [comment]
> Seems to me all of these can be done with `StdLib` already, there is `MemorySearch` there. `SUBSTR` opcode is available as well.

Are we assuming every developer is an expert of neo? Making it easier to use wont hurt.

File: src/Neo/SmartContract/Native/Regex.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+using System;
+
+namespace Neo.SmartContract.Native
+{
+    /// <summary>
+    /// A native contract library that provides useful regex functions.
+    /// By default, the maximum input length is 512 bytes.
+    /// We need to break regex functions into smaller pieces to calculate the gas cost.
+    /// </summary>
+    public sealed class Regex : NativeContract
```
Comment (shargon) [inline src/Neo/SmartContract/Native/Regex.cs:22]: I didn't like to add a native contract for this

File: src/Neo/SmartContract/Native/Regex.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+        /// <summary>
+        /// Get a substring of the given string
+        /// </summary>
+        /// <param name="str"> String to process</param>
+        /// <param name="startIndex"> Start index of the substring</param>
+        /// <param name="length">length of the substring</param>
+        /// <returns>The substring</returns>
+        [ContractMethod(CpuFee = 1 << 8)]
+        private static string Substring([MaxLength(MaxInputLength)] string str, int startIndex, int length)
```
Comment (shargon) [inline src/Neo/SmartContract/Native/Regex.cs:84]: We already have this https://github.com/neo-project/neo-vm/blob/ea03316ae34ca4bc7a30fd4b65dad74f1fed7e17/src/Neo.VM/ExecutionEngine.cs#L712

File: src/Neo/SmartContract/Native/Regex.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        }
+
+        /// <summary>
+        /// Determines whether the string instance contains the specified string.
+        /// </summary>
+        /// <param name="str">The string to process.</param>
+        /// <param name="value">The string to match.</param>
+        /// <returns>Whether [str] contains [value]</returns>
+        [ContractMethod(CpuFee = 1 << 8)]
+        private static bool Contains([MaxLength(MaxInputLength)] string str, [MaxLength(MaxInputLength)] string value)
```
Comment (shargon) [inline src/Neo/SmartContract/Native/Regex.cs:59]: IndexOf(x)!=-1

**Comment by Liaojinghui** [comment]
close this one. see https://github.com/neo-project/neo/pull/2927

**Pull request closed without merging**